File: Leetcode/js/9_palindromeNumber.js
/* Palindrome Number
Determine whether an integer is a palindrome. 
Do this without extra space.
*/

/**
 * @param {number} x
 * @return {boolean}
 */
var isPalindrome = function(x) {
    if(x<0){
        return false;
    }
    
    //find #of digits first
    var div = 1;
    while(x/div >= 10) {
        div *= 10;
    }
    
    while(x !== 0){
        var left = Math.floor(x/div);
        var right = x%10;
        //check
        if(left !== right) {
            return false;
        }
        //update
        //first, remove leftmost digit
        //then dividing by 10 remove rightmost digit
        x = Math.floor(x%div/10);
        //update div (removed 2 digits total)
        div /= 100;
    }
    return true;
    
};

File: Leetcode/java/9_palindromeNumber.java
/*
Determine whether an integer is a palindrome. Do this without extra space.

click to show spoilers.

Some hints:
Could negative integers be palindromes? (ie, -1)

If you are thinking of converting the integer to string, 
note the restriction of using extra space.

You could also try reversing an integer. 
However, if you have solved the problem "Reverse Integer", 
you know that the reversed integer might overflow. How would you handle such case?

There is a more generic way of solving this problem.
*/
class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println("Starting...");

        int input = 13221;
        boolean result = solution.isPalindrome(input);
        System.out.println("Result: " + result);
        
        System.out.println("DONE!");
    }
}

class Solution {
    public boolean isPalindrome(int x) {
        if(x<0){
            return false;
        }
        
        int tens = (int) (Math.log10(x));
        
        while(x>0 && tens>0) {
            int bigDivider = (int) Math.pow(10, tens);
            
            int left = x / bigDivider;
            int right = x % 10;
            
            if(left != right) {
                return false;
            }
            
            tens = tens-2;
            x = (x % bigDivider) / 10;
        }
        
        return true;
    }
}






